protocol/x/vault/keeper/withdraw_test.go

```diff
@@ -0,0 +1,430 @@
+	assettypes "github.com/dydxprotocol/v4-chain/protocol/x/assets/types"
+	clobtypes "github.com/dydxprotocol/v4-chain/protocol/x/clob/types"
+	perptypes "github.com/dydxprotocol/v4-chain/protocol/x/perpetuals/types"
+	pricestypes "github.com/dydxprotocol/v4-chain/protocol/x/prices/types"
+	satypes "github.com/dydxprotocol/v4-chain/protocol/x/subaccounts/types"
+	vaulttypes "github.com/dydxprotocol/v4-chain/protocol/x/vault/types"
+	"github.com/stretchr/testify/require"
+)
+
+func TestGetVaultWithdrawalSlippagePpm(t *testing.T) {
+	testVaultId := constants.Vault_Clob1
+	testClobPair := constants.ClobPair_Eth
+	testPerpetual := constants.EthUsd_20PercentInitial_10PercentMaintenance
+	testMarketParam := constants.TestMarketParams[1]
+	testMarketPrice := constants.TestMarketPrices[1]
+
+	tests := map[string]struct {
+		/* --- Setup --- */
+		// skew.
+		skewFactorPpm uint32
+		// spread.
+		spreadMinPpm      uint32
+		spreadBufferPpm   uint32
+		minPriceChangePpm uint32
+		// leverage.
+		assetQuoteQuantums   *big.Int
+		positionBaseQuantums *big.Int
+		// function input.
+		vaultId              vaulttypes.VaultId
+		withdrawalPortionPpm *big.Int
+		/* --- Expectations --- */
+		expectedSlippagePpm *big.Int
+		expectedErr         string
+	}{
+		"Success: leverage 0, skew 2, spread 0.003, withdraw 10%": {
+			skewFactorPpm:        2_000_000,
+			spreadMinPpm:         2_000,
+			spreadBufferPpm:      1_500,
+			minPriceChangePpm:    1_500,
+			assetQuoteQuantums:   big.NewInt(1_000_000_000), // 1,000 USDC
+			positionBaseQuantums: big.NewInt(0),
+			vaultId:              testVaultId,
+			withdrawalPortionPpm: big.NewInt(100_000), // 10%
+			// no slippage when leverage is 0.
+			expectedSlippagePpm: big.NewInt(0),
+		},
+		"Success: leverage 1.5, skew 2, spread 0.003, withdraw the smallest portion 0.0001%": {
+			skewFactorPpm:        2_000_000,
+			spreadMinPpm:         2_000,
+			spreadBufferPpm:      1_500,
+			minPriceChangePpm:    1_500,
+			assetQuoteQuantums:   big.NewInt(-1_000_000_000), // -1,000 USDC
+			positionBaseQuantums: big.NewInt(1_000_000_000),  // 1 ETH
+			vaultId:              testVaultId,
+			withdrawalPortionPpm: big.NewInt(1), // 0.0001%
+			// open_notional = 1_000_000_000 * 10^-9 * 3_000 * 10^6 = 3_000_000_000
+			// leverage = 3_000_000_000 / (-1_000_000_000 + 3_000_000_000) = 1.5
+			// posterior_leverage = 1.5 / (1 - 0.000001) = 1.5000015 ~= 1.500002
+			// integral
+			// = skew_antiderivative(skew_factor, posterior_leverage) - skew_antiderivative(skew_factor, leverage)
+			// = 2 * 1.500002^2 + 2^2 * 1.500002^3 / 3 - (2 * 1.5^2 + 2^2 * 1.5^3 / 3)
+			// = 9.000032 - 9
+			// = 0.000032
+			// average_skew = 0.000032 / (1.500002 - 1.5) = 16
+			// slippage = spread * (1 + average_skew) * leverage
+			// = 0.003 * (1 + 16) * 1.5
+			// = 0.0765
+			// slippage = min(0.0765, leverage * imf)
+			// = min(0.0765, 1.5 * 0.2) = 0.0765
+			expectedSlippagePpm: big.NewInt(76_500),
```

Comment:
Not sure if this is covered by [the min withdrawal amount convo](https://dydx-team.slack.com/archives/C06FEBLTXU7/p1726153828284799), but I get a different result (7.2%) for this situation:

```python
from decimal import Decimal

def skew_antiderivative(skew_factor, leverage):
    return skew_factor**2 / 3 * leverage**3 + skew_factor * leverage**2


leverage = Decimal('1.5')
skew = Decimal('2')
withdrawer_share = Decimal('0.000001')
half_spread_bps = Decimal('30')

posterior_leverage = (leverage / (1 - withdrawer_share))
integral_a = skew_antiderivative(skew, posterior_leverage)
integral_b = skew_antiderivative(skew, leverage) 
integral = integral_a - integral_b
average_skew = integral / (posterior_leverage - leverage)
fee_bps = half_spread_bps * (average_skew + 1) * leverage

{
    "posterior_leverage": posterior_leverage,
    "integral_a": integral_a,
    "integral_b": integral_b,
    "integral": integral,
    "average_skew": average_skew,
    "fee_bps": fee_bps
}
```

```python
{'posterior_leverage': Decimal('1.500001500001500001500001500'),
 'integral_a': Decimal('9.000022500040500063000089999'),
 'integral_b': Decimal('8.999999999999999999999999999'),
 'integral': Decimal('0.000022500040500063000090000'),
 'average_skew': Decimal('15.00001200001500001799995500'),
 'fee_bps': Decimal('720.0005400006750008099979750')}
```

Reply:
i this case, i think it's because `posterior_leverage` gets rounded up from `1.5000015` to `1.500002` as we operate with 6 decimal places, resulting in fee being 0.765 instead of 0.750